src/components/assignment-form/test-fields.jsx

```diff
@@ -0,0 +1,96 @@
+// @flow
+import Transition from 'react-motion-ui-pack';
+import { spring } from 'react-motion';
+import React, { Component } from 'react';
+import prefixer from 'utils/class-name-prefixer';
+import { connect } from 'react-redux';
+import type { State, Dispatch } from 'state/reducer';
+import { addTestFieldAction } from 'state/form';
+import IO from 'domain/io';
+import InputOutput from './input-output';
+
+class TestFields extends Component {
+
+  props: {
+    inputOutput: Array<IO>,
+    onAddFieldClick: () => void,
+    readOnly: boolean,
+  };
+
+  renderReadOnly() {
+    return (
+      <div>
+        {this.props.inputOutput.map((io: IO, index: number) =>
+          (<div key={io.hash()}>
+            {<InputOutput index={index} io={io} readOnly />}
+          </div>),
+        )}
+      </div>
+    );
+  }
+
+  renderEditable() {
+    return (
+      <div>
+        <div className={prefixer('io-component')}>
+          <Transition
+            appear={{
+              opacity: 0,
+              height: 0,
+              translateY: 80,
+              translateX: 0,
+            }}
+            enter={{
+              overflow: 'hidden',
+              height: 64,
+              opacity: 1,
+              translateX: 0,
+              translateY: spring(0, { stiffness: 120, damping: 15 }),
+            }}
+            leave={{ opacity: 0, height: 0 }}
+          >
+            {this.props.inputOutput.map((io: IO, index: number) =>
+              (<div key={io.hash()}>
+                {<InputOutput index={index} io={io} />}
+              </div>),
+            )}
+          </Transition>
+        </div>
+        <button
+          type="button"
+          className={prefixer('add-field')}
+          onClick={(e) => { e.preventDefault(); this.props.onAddFieldClick(); }}
+        >
+          + Lisää kenttä
+      </button>
+      </div>
+    );
+  }
+
+  render() {
+    return (
+      <div className={prefixer('form-component')}>
+        <div className={prefixer('instructions')}>
+          Testit
+        </div>
+        {this.props.readOnly ? this.renderReadOnly() : this.renderEditable()}
```

Comment:
If editable and read-only variants are that different does it make sense to extract these into their own components?